src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/SemanticTokens.cs

```diff
@@ -77,4 +81,135 @@ internal partial class RazorCustomMessageTarget
+        }
+
+        var (synchronized, csharpDoc) = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync<CSharpVirtualDocumentSnapshot>(
+            _documentManager, (int)semanticTokensParams.RequiredHostDocumentVersion, semanticTokensParams.TextDocument, cancellationToken);
+
+        if (csharpDoc is null)
+        {
+            return null;
+        }
+
+        if (!synchronized)
+        {
+            // If we're unable to synchronize we won't produce useful results, but we have to indicate
+            // it's due to out of sync by providing the old version
+            return new ProvideSemanticTokensResponse(tokens: null, hostDocumentSyncVersion: -1);
+        }
+
+        // Ensure the C# ranges are sorted
+        Array.Sort(semanticTokensParams.Ranges, static (r1, r2) => r1.CompareTo(r2));
+
+        semanticTokensParams.TextDocument.Uri = csharpDoc.Uri;
+        var requestTasks = new List<Task<ReinvocationResponse<VSSemanticTokensResponse>?>>(semanticTokensParams.Ranges.Length);
+        var textBuffer = csharpDoc.Snapshot.TextBuffer;
+        var languageServerName = RazorLSPConstants.RazorCSharpLanguageServerName;
+        var lspMethodName = Methods.TextDocumentSemanticTokensRangeName;
+
+        foreach (var range in semanticTokensParams.Ranges)
+        {
+            var newParams = new SemanticTokensRangeParams
+            {
+                TextDocument = semanticTokensParams.TextDocument,
+                PartialResultToken = semanticTokensParams.PartialResultToken,
+                Range = range,
+            };
+
+            using var _ = _telemetryReporter.TrackLspRequest(lspMethodName, languageServerName, semanticTokensParams.CorrelationId);
+            var task = _requestInvoker.ReinvokeRequestOnServerAsync<SemanticTokensRangeParams, VSSemanticTokensResponse>(
+                textBuffer,
+                lspMethodName,
+                languageServerName,
+                newParams,
+                cancellationToken);
+            requestTasks.Add(task);
+        }
+
+        var results = await Task.WhenAll(requestTasks).ConfigureAwait(false);
+        var nonEmptyResults = results.Select(r => r?.Response?.Data).WithoutNull().ToArray();
+
+        if (nonEmptyResults.Length != semanticTokensParams.Ranges.Length)
+        {
+            // Weren't able to re-invoke C# semantic tokens but we have to indicate it's due to out of sync by providing the old version
+            return new ProvideSemanticTokensResponse(tokens: null, hostDocumentSyncVersion: csharpDoc.HostDocumentSyncVersion);
+        }
+
+        var data = StitchSemanticTokenResponsesTogether(nonEmptyResults);
+
+        var response = new ProvideSemanticTokensResponse(data, semanticTokensParams.RequiredHostDocumentVersion);
+
+        return response;
+    }
+
+    private int[] StitchSemanticTokenResponsesTogether(int[][] responseData)
+    {
+        var count = responseData.Sum(r => r.Length);
```

Comment:
Isn't this just `responseData.Length * 5`?

Reply:
No, each of the passed in array is of a different length, so we need to make a pass through to determine what the stitched array's length should be.